boards/earlgrey-nexysvideo/src/tests.rs

```diff
@@ -0,0 +1,96 @@
+use crate::debug::IoWrite;
+use core::fmt::Write;
+use core::panic::PanicInfo;
+use kernel::Chip;
+
+use crate::CHIP;
+
+extern "C" {
+    pub(crate) fn semihost_command(command: usize, arg0: usize, arg1: usize) -> !;
+}
+
+#[cfg(test)]
+pub(crate) fn test_runner(tests: &[&dyn Fn()]) {
+    for test in tests {
+        test();
+    }
+
+    // Make sure all the messages are printed
+    for _ in 0..200000 {
+        rv32i::support::nop();
+    }
+
+    // Exit QEMU with a return code of 0
+    unsafe {
+        semihost_command(0x18, 0x20026, 0);
+    }
+}
+
+struct Writer<'a> {
+    uart: earlgrey::uart::Uart<'a>,
+}
+
+// This creates a second UART peripheral.
+// This should ONLY be used by test cases running on QEMU.
+static mut WRITER: Writer = Writer {
+    uart: earlgrey::uart::Uart::new(
+        earlgrey::uart::UART0_BASE,
+        earlgrey::chip_config::CONFIG.peripheral_freq,
+    ),
+};
```

Comment:
It seems possible for there to be weird behavior in the tests if they call any code in the kernel containing debug statements, once there are two writers. I assume the issue is that you need a synchronous writer here?

Reply:
Yes, there could be. The problem is because we don't go into the kernel loop we can't use the usual `debug!`. That actually is something we could maybe improve in the future, by maybe running the tests as part of the loop (or replacing the loop with a custom test loop)